MS5xxx: Refactor into common driver for all members of the sensor family

kubark42 opened the pull request:
Measurement Specialties has several sensors that are built around the same circuitry with different trim values. The upshot is that the driver is almost identical for I2C and SPI, and for all sensors in this family. At the least, this includes the:
- MS5611
- MS5637
- MS5803-01
- MS5803-02
- MS5803-05
- MS5803-07
- MS5803-14
- MS5803-30

The only differences between the above chips are the scalar multipliers and divisors for calculating the second-order compensated values.

kubark42 commented:
Driver is now tested and works. However, as a result of a phone conversation with a Measurement Specialties application engineer, I'm going to be refactoring this into the existing MS5611 driver. He confirmed that this is identical circuitry with the only differentiator being the package.

Each sensor range has slight different math, but those are the only differences in the code.

kubark42 committed:
MS5xxx: Rename all MS5611 resources to MS5xxx

***BREAKS COMPILATION***

kubark42 committed:
BMP085: Fix minor typo

This driver is for the BMP085, not the MS5611.

kubark42 committed:
MS5xxx baro: Merged SPI and I2C support into one file

Includes all targets as well.

kubark42 committed:
MS5xxx baro: Declare the MS5xxx chip model

All boards use MS5611 at this time.

kubark42 committed:
MS5xxx baro: Require specifying the MS5xxx chip model

Otherwise, the function returns an error and the board will go into panic.

Also, change the error value returned.

kubark42 committed:
MS5xxx baro: Simplify pressure conversion

There was no need to divide both values by 1000 before taking their ratio.

kubark42 committed:
MS5xxx baro: Use different math for different sensors

The math for each sensor in the MS5xxx line is slightly different. So far,
it is only a difference in the scalar value of the multipliers and
divisors. This likely means that the section can be further refactored by
only handling the differences in the scalars separately.

Also moved the conversion math into a separate function. This function is
free of any PiOS calls and so can be checked with unit tests. This will be
helpful for refactoring any of the math

kubark42 commented:
@peabody124 I just pushed the refactoring. I'd appreciate help in testing any of the below.
- [ ] AQ32
- [x] Colibri
- [ ] Naze32
- [x] Quanton
- [ ] RevoMini
- [x] Sparky
- [x] Sparky 2

kubark42 committed:
MS5XXX: Replace (almost) all instances of MS5611 by MS5XXX

There are still a few specific instances of MS5611 where it is referring to
the chip installed on specific targets.

kubark42 committed:
SparkyBGC: Remove vestigal MS5611 barometric sensor compilation

The sensor is not installed on the SparkyBGC.

kubark42 commented:
I forgot to say that this adds support for the MS5637, MS5803-01, and MS5803-02, i.e. not all the chips in the MS5xxx family.

peabody124 commented:
I should be able to test a number of those targets by this weekend

kubark42 commented:
Thanks. I've tested the MS5637 I2C version and MS5803-02 SPI version as well.

mlyle commented on flight/PiOS/Common/pios_ms5xxx.c:
```diff
@@ -0,0 +1,698 @@
+	int32_t rc = 0;
+
+	switch (dev->pios_bus_type) {
+	case PIOS_SPI_BUS:
```
The issue with this structure is that it ends up including both the SPI and the I2C code irrespective of which is being used on the target.

kubark42 replied:
It will increase code complexity to conditionally include I2C and SPI, whereas the existing approach seems to work fine without the conditional includes. Is there an advantage to what you're suggesting beyond minimizing impact on F1 processors?

mlyle replied:
Sparky1 is nearly out of space, too, and is a really nice target.

mlyle commented on flight/PiOS/Common/pios_ms5xxx.c:
```diff
@@ -0,0 +1,698 @@
+ */
+static int32_t PIOS_MS5XXX_ClaimSPIBus(void)
+{
+#if defined(PIOS_INCLUDE_SPI)
```
This particular group is wrapped in PIOS_INCLUDE_SPI but other stuff below uses SPI unconditionally.

peabody124 replied:
Probably worth duplicating this include down around L497 or removing it here.

kubark42 replied:
These #ifdefs were included because of compile-time failures when `PIOS_INCLUDE_SPI` wasn't set. I will look into why the transfer byte and transfer block sections (~L497) don't cause compiler failures (because I kinda think they should).

kubark42 commented on flight/PiOS/Common/pios_ms5xxx.c:
```diff
@@ -0,0 +1,700 @@
+	PIOS_SENSORS_Register(PIOS_SENSOR_BARO, dev->queue);
+
+	dev->task = PIOS_Thread_Create(
+			PIOS_MS5XXX_Task, "pios_ms5xxx", MS5XXX_TASK_STACK_BYTES, NULL, MS5XXX_TASK_PRIORITY);
```
Should this task be added to the task manager UAVObject? I don't really see any downsides to this, but it wasn't done before so I didn't add it.

kubark42 commented on flight/PiOS/Common/pios_ms5xxx.c:
```diff
@@ -0,0 +1,700 @@
+
+/* Private constants */
+#define MS5XXX_TASK_PRIORITY	PIOS_THREAD_PRIO_HIGHEST
+#define MS5XXX_TASK_STACK_BYTES	512
```
Missing space.

peabody124 commented:
Tested Quanton and Sparky1

peabody124 commented:
Tested Colibri

peabody124 commented on flight/PiOS/Common/pios_ms5xxx.c:
```diff
@@ -0,0 +1,700 @@
+	dev->cfg = cfg;
+
+	// Check that a model of the chip was chosen
+	if (cfg->pios_ms5xxx_model == 0) {
```
Is there any way to confirm the device we are talking to matches the model?

kubark42 replied:
Alas, not. In Measurement Specialties's chip docs, I didn't see a WHOAMI or something that would be unique to each different model in the series. That's a real shame, because you will get the wrong absolute altitude if you use the wrong driver math, but your relative altitude changes will still be sort-of correct. (Ask me how I know!)

peabody124 commented:
Aside from comments from @mlyle regarding the conditionals (I don't feel strongly) LGTM and good to merge. Nice improvement!

Commit:
MS5XXX: Removed all references to PIOS_INCLUDE_MS5XXX_SPI

This was no longer necessary as the code now handles both I2C and SPI in
the same routines.

Commit:
MS5XXX: Fixes to compile safely with PIOS include flags when -O0

Compiling would fail when compiling with -O0 because the compiler would no
longer automatically eliminate impossible-to-reach pathways. The upshot was
that functions that included calls to SPI or I2C would be unable to link.

kubark42 commented:
The compiler removes unused sections and so the additional #includes aren't necessary for normal use as the outputted `-Os` binaries are identical, but we still need all the conditional includes for compiling `-O0`. I added sufficient `PIOS_INCLUDE_SPI` and `PIOS_INCLUDE_I2C` to ensure this would compile correctly with `-O0` no matter the options chosen. 

As an aside, it looks like we might have breakage elsewhere in the project if we compile `-O0`. I know that's unlikely to do in practice, but it does seem we should be testing this pathway in Jenkins because I think it's pretty reasonable to expect `-O0` to compile successfully.
